projects/core/src/config/config-initializer/config-initializer.service.ts

```diff
@@ -0,0 +1,117 @@
+import { Inject, Injectable, isDevMode, Optional } from '@angular/core';
+import { CONFIG_INITIALIZER, ConfigInitializer } from './config-initializer';
+import { Config } from '../config.module';
+import { BehaviorSubject } from 'rxjs';
+import { filter, mapTo, take } from 'rxjs/operators';
+import { deepMerge } from '../utils/deep-merge';
+
+@Injectable({
+  providedIn: 'root',
+})
+export class ConfigInitializerService {
+  constructor(
+    @Inject(Config) protected config: any,
+    @Optional()
+    @Inject(CONFIG_INITIALIZER)
+    protected initializers: ConfigInitializer[]
+  ) {}
+
+  protected ongoingScopes$ = new BehaviorSubject<string[]>(undefined);
+
+  async getStableConfig(scopes: string[] = []): Promise<any> {
+    if (!this.config.initializing) {
+      return this.config;
+    }
+    return this.ongoingScopes$
+      .pipe(
+        filter(
+          ongoingScopes => ongoingScopes && this.areReady(ongoingScopes, scopes)
+        ),
+        take(1),
+        mapTo(this.config)
+      )
+      .toPromise();
+  }
+
+  protected finishScope(scopes: string[]) {
+    const newScopes = [...this.ongoingScopes$.value];
+    for (const scope of scopes) {
+      newScopes.splice(newScopes.indexOf(scope), 1);
+    }
+    this.ongoingScopes$.next(newScopes);
+  }
+
+  protected areReady(ongoing: string[], waitingFor: string[]): boolean {
+    if (!waitingFor.length) {
+      return !ongoing.length;
+    }
+    for (const waitingScope of waitingFor) {
+      if (!this.isReady(ongoing, waitingScope)) {
+        return false;
+      }
+    }
+    return true;
+  }
+
+  protected isReady(ongoing: string[], waitingScope: string): boolean {
+    for (const scope of ongoing) {
+      if (waitingScope.includes(scope)) {
+        return false;
+      }
+    }
+    return true;
+  }
+
+  /**
+   * @internal
+   */
+  async initialize() {
+    if (this.ongoingScopes$.value) {
+      // guard against multiple initialization calls
+      return;
+    }
+
+    if (!(this.initializers && this.initializers.length)) {
+      deepMerge(this.config, { initializing: false });
+      this.ongoingScopes$.next([]);
+      return;
+    }
+
+    const ongoingScopes = ['initializing'];
+
+    const asyncConfigs = [];
+
+    for (const initializer of this.initializers) {
+      if (isDevMode() && !(initializer.scopes && initializer.scopes.length)) {
+        console.error('ConfigInitializer should provide scope!');
```

Comment:
What about replacing `ConfigInitializer` to the name of the token `CONFIG_INITIALIZER` in both messages - for discoverability?

Reply:
👍